content/entry-hash/index.md

```diff
@@ -0,0 +1,239 @@
+has affected the result of the entry hash even though it is (a) not stable and
+(b) not part of the real identity of an entry.
+
+Finally, the current implementation relies on JSON and on the attribute names
+serialised as strings.
+
+This RFC aims to provide an algorithm that doesn't rely on specific attribute
+names (e.g. `entry-number` vs `number`, `item-hash` vs `itemHash`, etc) which
+supports a better decoupling between the data model and the REST API (we don't
+have plans for it but a hypothetical GraphQL interface wouldn't work with the
+current attribute names and yet conceptually an entry is the same).
+
+Note that the item hash algorithm is not affected by this RFC.
+
+## Explanation
+
+This section describes the algorithm twice. First in a high level and then a
+more step by step with intermediate values.
+
+For reference, this is the abstract definition of an entry:
+
+```elm
+type Entry =
+  { number : Integer
+  , key : ID
+  , timestamp : Timestamp
+  , item : Set Hash
+  }
+```
+
+### The algorithm
+
+When this algorithm operates on _hashes_ (e.g. tag, concatenate) it is done on
+the byte level, not the hexadecimal string representation that the latter
+example shows as partial representations.
+
+1. Let _number_ be the string representation of the entry number.
+2. Let _key_ be the string representation of the entry key.
+3. Let _timestamp_ be the string representation of the entry timestamp.
+4. Let _items_ be the set of item hashes as bytes. Note that
+   the item hashes must not have the prefix (i.e. `sha-256:`) when transformed to
+   bytes.
+5. Let _hashList_ be an empty list.
+6. Apply the _hashValue_ function to _number_ tagged with `i` (Integer). And
+   append the result to _hashList_.
+7. Apply the _hashValue_ function to _key_ tagged with `u` (String). And
+   append the result to _hashList_.
+8. Apply the _hashValue_ function to _timestamp_ tagged with `t` (Timestamp). And
+   append the result to _hashList_.
+9. Let _itemsBytes_ be
+10. For each value _item_ in the _items_ set:
+   1. Apply the _hashValue_ function to _item_ tagged with `r` (Hash) and
+      append the result to _itemsBytes_.
+11. [Sort the elements](#sorting) of _itemsBytes_, concatenate them and
+12. Apply the _hashValue_ function tagged with `s` (Set). And append the result
+    to _hashList_.
+13. Concatenate the elements of _hashList_ in order (i.e. `[numberHash,
+    keyHash, timestampHash, itemsHash]`), tag it with `l` (List) and hash the result.
+
+#### Sorting
+
+The sorting algorithm for a set of hashes is done by comparing the list of
+bytes one by one. For example, given a set `["foo", "bar"]` you'll get the
+folllowing byte lists after hashing them as unicode:
+
+```elm
+[ [166,166,229,231,131,195,99,205,149,105,62,193,137,194,104,35,21,217,86,134,147,151,115,134,121,181,99,5,242,9,80,56]
+, [227,3,206,11,208,244,193,253,254,76,193,232,55,215,57,18,65,226,224,71,223,16,250,97,1,115,61,193,32,103,93,254]
+]
+```
```

Comment:
To get this result you must prepend "foo" and "bar" with 'u'. This sorting algorithm is used in step 11 where the individual components of _itemBytes_ have been preprended with 'r'. Is it worth changing this example to make it clear that 'r' is used to tag and not 'u'?

Reply:
Or just make it clear that in this example the strings have been tagged with `u` though the bytes may have come from any tag and any hash.

Reply:
I think I might have changed my mind about this, it does specify "hashing them as unicode" which implied the tag 'u' is used. Ignore me.